Initial Github Actions CI

MichaelOultram opened the pull request:
I'm not sure if you want to enable all these jobs, but I have got them all passing

* Ensure code is formatted correctly with `cargo fmt --all`
* Ensure code passes the clippy lints (`cargo clippy --workspace`)
* ~Ensure dependencies are listed in order (`cargo sort --workspace`) (requires [cargo-sort](https://crates.io/crates/cargo-sort))~
* Check dependencies for advisories, bans, licenses and sources using `cargo deny check` (requires [cargo-deny](https://github.com/EmbarkStudios/cargo-deny))
* Ensure application builds, and tests pass on stable, nightly, and MSRV rust versions
* When running tests under nightly, produce a coverage output (currently uploaded as an artifact, could be better...).

Note: This uses [cargo nextest](https://nexte.st/) instead of the default `cargo test` to run tests.

MichaelOultram commented:
It appears you will need to approve the PR before github actions will be able to run here: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

Horus645 commented:
This is very cool, thanks!

It will take me a little while to review since I am not very familiar with GitHub Actions though.

MichaelOultram commented:
> In any case, can you please delete the sort check? I prefer to organize dependencies logically, rather than alphabetically.

Understandable, I have removed cargo-sort and rebased the branch. 

However, I'm getting some weird issue with `cargo fmt`. It is failing in this PR, but succeeding in my [test PR](https://github.com/MichaelOultram/swww/pull/3). As far as I can tell the versions of rustfmt are the same, so I really don't know what is causing this. I'll convert this PR back to a draft until I can figure this out.

MichaelOultram commented:
Ah, I figured it out. I did not realize that the file `cargo fmt` was complaining about [was edited today](https://github.com/LGFae/swww/commit/1eecca7d83026ab3bf5555d3f43310a699e22d2a). My fork didn't have that change at the time, and so I got different results. All fixed and ready to review again.

max-ishere commented:
YO yo before you merge I have somethng im working on like rn

max-ishere commented:
I will send a full response in a few minutes

max-ishere commented:
Its not drastic difference, unless its a noop like for --bins, but its a good idea either way to speed up workflows

max-ishere commented:
Also these docs for reference: https://doc.rust-lang.org/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci

max-ishere commented:
@MichaelOultram how did you by chance decide to work on it at the same time as I did, thats a weird coincidence /reh.

MichaelOultram commented:
> how did you by chance decide to work on it at the same time as I did, thats a weird coincidence

I guess we both love the advantages of having CI enabled on a repo. :D

I opted to use [sccache](https://github.com/mozilla/sccache/) so dependencies won't be rebuilt each time. The cache stats are shown on the post run step of each build.
```
Compile requests                    206
Compile requests executed           166
Cache hits                          162
Cache hits (C/C++)                    4
Cache hits (Rust)                   158
Cache misses                          2
Cache misses (Rust)                   2
```

I could add caching of the cargo folder also, but my suspicion is that it wouldn't save any time. The cache either has to be downloaded from GitHub, or the original files are downloaded from cargo. Either way there is a download involved.

Also, I have a concern that the cargo folder would increase in size over time. E.g. If one commit removes/updates a dependency, I believe it would download the existing cache, build (pulling the new dependency, but not removing the old one), and then upload the new cache.

max-ishere commented:
> Also, I have a concern that the cargo folder would increase in size over time. E

No, it makes a new cache if `Cargo.lock` changes. Actually my snippet ignores any sub-dir `Cargo.lock`s, which can be fixed with `${{ hashFiles('**/Cargo.lock') }}`. I only have one package in my repo tho, so its not a problem rn.

max-ishere commented:
Here's my caches, but I just recently added more folders for caching so its not very fair.
https://github.com/max-ishere/grub-theme/actions/caches?query=sort%3Aaccessed-desc

max-ishere commented:
Also I doubt it will ever reach the 10gb limit
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

MichaelOultram commented:
In your example, you have included restore-keys of
```
restore-keys: |
          cargo-build-${{ hashFiles('Cargo.lock') }}
          cargo-build-
          cargo-
```
So it will try to fetch the cache related to the hash of the current lock file, but if that doesn't exist it will fetch the latest cache with the prefix `cargo-build-` (i.e. the old cache). Then it will build and upload as `cargo-build-newhash`. [The example in the docs](https://github.com/actions/cache/blob/main/examples.md#rust---cargo) doesn't include restore-keys, possibly for this reason.

My concern about unnecessary size is that it'll take longer to download. But if we remove the restore-keys, I believe the size will be sorted. I'll try to get this in tonight.

MichaelOultram commented:
Ah, good spot. So my initial thought is wrong, the cache won't grow in size, it just won't be able to save a new cache if it's restored an old cache. Removing the restore-keys should still fix this.

max-ishere commented:
Yeah, well its still recompiling...

max-ishere commented:
I dont understand... the time improves but it still recompiles everything

max-ishere commented:
```
gh cache list

Showing 16 of 16 caches in max-ishere/grub-theme

ID  KEY                                                                           SIZE        CREATED             ACCESSED          
77  cargo-doc-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e    105.72 MiB  about 10 hours ago  about 10 hours ago
67  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e  93.70 MiB   about 10 hours ago  about 10 hours ago
66  cargo-test-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   93.78 MiB   about 10 hours ago  about 10 hours ago
65  cargo-lint-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   93.70 MiB   about 10 hours ago  about 10 hours ago
36  cargo-lint-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   1.11 KiB    about 6 days ago    about 10 hours ago
62  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e  93.70 MiB   about 10 hours ago  about 10 hours ago
37  cargo-test-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   1.10 KiB    about 6 days ago    about 10 hours ago
61  cargo-doc-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e    12.35 MiB   about 6 days ago    about 5 days ago
35  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e  1.10 KiB    about 6 days ago    about 5 days ago
50  cargo-test-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   12.35 MiB   about 6 days ago    about 6 days ago
49  cargo-lint-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   12.35 MiB   about 6 days ago    about 6 days ago
48  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e  12.35 MiB   about 6 days ago    about 6 days ago
47  cargo-doc-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e    12.35 MiB   about 6 days ago    about 6 days ago
29  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e  1.11 KiB    about 12 days ago   about 6 days ago
34  cargo-lint-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   1.29 KiB    about 12 days ago   about 6 days ago
28  cargo-test-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505405f2e   1.10 KiB    about 12 days ago   about 6 days ago
```
```
gh cache delete --all
✓ Deleted 16 caches from max-ishere/grub-theme
```
Re-run all workflows
```
gh cache list

Showing 4 of 4 caches in max-ishere/grub-theme

ID  KEY                                                                       SIZE        CREATED                 ACCESSED              
86  cargo-test-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505...  93.78 MiB   less than a minute ago  less than a minute ago
82  cargo-lint-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd57505...  93.70 MiB   less than a minute ago  less than a minute ago
81  cargo-doc-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd575054...  105.72 MiB  less than a minute ago  less than a minute ago
78  cargo-build-b74815bc8fca3c601f720bbe8dc4b27b95de751086213d49bd9bd5750...  93.70 MiB   about 1 minute ago      about 1 minute ago
```
Timing results (i dont know how its ordered):

[clippy](https://github.com/max-ishere/grub-theme/actions/runs/7893104007/usage)
[gh-pages-cargo-doc](https://github.com/max-ishere/grub-theme/actions/runs/7893104005/usage)
[test](https://github.com/max-ishere/grub-theme/actions/runs/7893104003/usage)
[build](https://github.com/max-ishere/grub-theme/actions/runs/7893104002/usage)

max-ishere commented:
I mean not all of them are faster for some reason

max-ishere commented:
Oh ok, it seems they are ordered by run attempt

max-ishere commented:
I think im gonna try adding sccache to my repo to see if I can avoid the recompiling that cargo does all the time

max-ishere commented:
Also it seems its not caching https://github.com/LGFae/swww/actions/runs/7892740838/job/21539892335 its still running some compilation, unless that is just visual and its all handled by sccache.. Hmm worth checking.

MichaelOultram commented:
I believe it is just visual. The compiling line comes from cargo, and then it invokes rustic (wrapped with sccache). At this point, sccache can either pull the cache if it exists, or run the original rustc command if not. To see how successful sccache is at pulling from the cache, you need to look at the post action on a run to see how many hits/misses there are.

max-ishere commented:
Oh, ok, this looks pretty good actually
[Post Run sccache](https://github.com/LGFae/swww/actions/runs/7892740838/job/21539892540)
```
Post job cleanup.
/opt/hostedtoolcache/sccache/0.7.7/x64/sccache --show-stats
Compile requests                    206
Compile requests executed           166
Cache hits                          162
Cache hits (C/C++)                    4
Cache hits (Rust)                   158
Cache misses                          2
Cache misses (Rust)                   2
Cache timeouts                        0
Cache read errors                     0
Forced recaches                       0
Cache write errors                    0
Compilation failures                  2
Cache errors                          0
Non-cacheable compilations            0
Non-cacheable calls                  40
Non-compilation calls                 0
Unsupported compiler calls            0
Average cache write               0.622 s
Average compiler                  0.651 s
Average cache read hit            0.292 s
Failed distributed compilations       0

Non-cacheable reasons:
crate-type                           38
-                                     2
```

max-ishere commented:
> Ah, good spot. So my initial thought is wrong, the cache won't grow in size, it just won't be able to save a new cache if it's restored an old cache. Removing the restore-keys should still fix this.

Hmm that's potentially a problem because if you get a cache hit, you are risking restoring a cache with unneeded crates... So maybe before cache-save runs I should run some sort of cache cleaner in cargo. idk if it will work with the $CARGO_HOME dir tho...

max-ishere commented:
I kinda like the download cache I think it improves the speed. But its not something I have the energy to compare.

max-ishere commented:
Alright I think I'm done here, good to learn about sccache. I dont think I have much more to do here anymore.

LGFae commented:
Thanks a ton, guys. This will be a huge help going forward!
